pkg/gadget-collection/gadgets/traceloop/gadget.go

```diff
@@ -52,138 +69,426 @@ some differences:
+	// Then, the CLI receives this big line, parses it and prints each event.
+	// A proper solution would be to develop a specific "output" (neither status
+	// nor stream) for traceloop, this is let as TODO.
+	t.helpers.PublishEvent(traceName, string(r))
 
-	timeout := time.After(2 * time.Second)
+	trace.Status.State = gadgetv1alpha1.TraceStateCompleted
+}
 
-	done := make(chan struct{})
-	go func() {
-		t.cmd.Process.Wait()
-		done <- struct{}{}
-	}()
+func (t *Trace) Delete(trace *gadgetv1alpha1.Trace) {
+	if trace.Spec.OutputMode != gadgetv1alpha1.TraceOutputModeStatus {
+		trace.Status.OperationError = fmt.Sprintf("\"delete\" operation can only be used with %q trace while %q was given", gadgetv1alpha1.TraceOutputModeStatus, trace.Spec.OutputMode)
 
-	select {
-	case <-timeout:
-		t.cmd.Process.Kill()
-	case <-done:
+		return
 	}
 
-	t.started = false
+	traceUnique.Lock()
+	if traceUnique.tracer == nil {
+		traceUnique.Unlock()
+
+		trace.Status.OperationError = "Traceloop tracer is nil"
+
+		return
+	}
+	traceUnique.Unlock()
+
+	containerID := trace.Spec.Parameters["containerID"]
+	_, ok := t.containerIDs[containerID]
+	if !ok {
+		ids := make([]string, len(t.containerIDs))
+		i := 0
+		for id := range t.containerIDs {
+			ids[i] = id
+			i++
+		}
+
+		trace.Status.OperationError = fmt.Sprintf("%q is not a valid ID for this trace, valid IDs are: %v", containerID, strings.Join(ids, ","))
+
+		return
+	}
+
+	traceUnique.Lock()
+	// First, we need to remove the perf buffer.
+	err := traceUnique.tracer.Delete(containerID)
+	traceUnique.Unlock()
+	if err != nil {
+		trace.Status.OperationError = fmt.Sprintf("Failed to delete perf buffer: %s", err)
+
+		return
+	}
+
+	// Then, we remove containerID from the map.
+	delete(t.containerIDs, containerID)
+
+	// Finally, we need to update the trace output.
+	var infos []types.TraceloopInfo
+	err = json.Unmarshal([]byte(t.trace.Status.Output), &infos)
+	if err != nil {
+		trace.Status.OperationError = fmt.Sprintf("failed to unmarshal output: %s", err)
+
+		return
+	}
+
+	newInfos := make([]types.TraceloopInfo, len(infos)-1)
+	i := 0
+	for _, info := range infos {
+		// We copy all the current information expect the one corresponding to the
+		// container we removed.
+		if info.ContainerID == containerID {
+			continue
+		}
+
+		newInfos[i] = info
+		i++
+	}
+
+	output, err := json.Marshal(newInfos)
+	if err != nil {
+		trace.Status.OperationError = fmt.Sprintf("failed to marshal new infos: %s", err)
+
+		return
+	}
+
+	traceBeforePatch := t.trace.DeepCopy()
+	t.trace.Status.Output = string(output)
+	patch := client.MergeFrom(traceBeforePatch)
 
-	return nil
+	// We also need to manually patch the trace CRD to have our modifications be
+	// taken into account.
+	err = t.client.Status().Patch(context.TODO(), t.trace, patch)
+	if err != nil {
+		trace.Status.OperationError = fmt.Sprintf("failed to patch trace %q output: %s", trace.Name, err)
```

Comment:
This needs to be a `log.Errorf`: changes to `trace.Status.OperationError` are not visible if it failed to patch the trace CR.